frontend/src/components/AIVisualization.jsx

```diff
@@ -0,0 +1,174 @@
+import { useState, useEffect } from "react";
+import "./AIVisualization.css";
+import { onAuthStateChanged } from "firebase/auth";
+import { auth, generateContent } from "../firebase";
+import { useRef } from "react";
+
+function AIVisualization(props) {
+  let ref = useRef();
+  const renderCircleNumber = 16;
+  const findTheta = (w, h) => Math.atan(w / h) / (Math.PI / 180); //degrees
+
+  const [mousePos, setMousePosition] = useState({
+    x: null,
+    y: null,
+  });
+  const [currI, setCurrI] = useState(0);
+
+  const getPixelRatio = (context) => {
+    var backingStore =
+      context.backingStorePixelRatio ||
+      context.webkitBackingStorePixelRatio ||
+      context.mozBackingStorePixelRatio ||
+      context.msBackingStorePixelRatio ||
+      context.oBackingStorePixelRatio ||
+      context.backingStorePixelRatio ||
+      1;
+
+    return (window.devicePixelRatio || 1) / backingStore;
+  };
+
+  useEffect(() => {
+    let canvas = ref.current;
+    let context = canvas.getContext("2d");
+
+    let ratio = getPixelRatio(context);
+    let width = getComputedStyle(canvas).getPropertyValue("width").slice(0, -2);
+    let height = getComputedStyle(canvas)
+      .getPropertyValue("height")
+      .slice(0, -2);
+
+    canvas.width = width * ratio;
+    canvas.height = height * ratio;
+    canvas.style.width = `${width}px`;
+
+    const updateMousePosition = (ev) => {
+      if (ref) {
+        setMousePosition({
+          x: Math.floor(
+            ((ev.clientX - ref?.current?.getBoundingClientRect().left) * 100) /
+              width
+          ),
+          y: Math.floor(
+            ((ev.clientY - ref?.current?.getBoundingClientRect().top) * 100) /
+              height
+          ),
+        });
+      }
+    };
+
+    window.addEventListener("mousemove", updateMousePosition);
+
+    const circle = (x, y, r, color, a, w, border) => {
+      context.fillStyle = color;
+      context.globalAlpha = a;
+
+      context.beginPath();
+
+      context.arc(
+        canvas.width * (x / 100),
+        canvas.height * (y / 100),
+        (canvas.width / 2) * (r / 100),
+        0,
+        2 * Math.PI
+      );
+
+      context.fill();
+
+      if (border) {
+        context.lineWidth = w;
+        context.stroke();
+      }
+      context.fillStyle = "black";
+      context.globalAlpha = 1;
+      context.lineWidth = 0;
+      context.closePath();
+    };
+
+    circle(50, 50, 50, "white");
+
+    let requestId,
+      i = currI;
+    let forceArray = Array(renderCircleNumber).fill([0, 0]);
+    const render = () => {
```

Comment:
render() function is a little bit long, can we break it down and add more comments?

Also maybe try to use useCallback() to wrap render() for better performance.

Reply:
You are right, lots of Math that gets abstracted in the code. I can do that. I will also look into useCallback! Documentation makes it sound pretty helpful.